record-encryption: document KEK rotation implications and config (#1411)
kroxylicious/kroxylicious: docs/modules/record-encryption/con-record-encryption-overview.adoc

Key rotation involves periodically replacing cryptographic keys with new ones and is considered a best practice in cryptography.

The filter allows the rotation of Key Encryption Keys (KEKs) within the Key Management System (KMS). 
When a KEK is rotated, the new key material is eventually used for newly produced records. Existing records, encrypted with older KEK versions, remain decryptable as long as the previous KEK versions are still in the KMS.

> When the KEK is rotated in the external KMS, it will take up to 2 hours (by default) before all records produced by the filter

Reviewer: I almost wonder if `2 hours` is too short given the cost implications of that decision. Maybe 6 or 12 would make for a better default.

Author: I realised I need to document the refresh config option too, it's more like 1hour effectively. 2hours is when it's eligible for eviction. We can bump the defaults and doc together in another PR if it's too small.

Reviewer 2: I think the defaults are good enough for now.

contain a DEK encrypted with the new key material. This is because existing encrypted DEKs are used for a configurable

> amount of time after creation, the Filter caches the encrypted DEK.

Reviewer: Should we also document the nuclear option of  a (rolling) restart of instances if people are responding to a key compromise?

Author: done

Key rotation involves periodically replacing cryptographic keys with new ones and is considered a best practice in cryptography.

The filter allows the rotation of Key Encryption Keys (KEKs) within the Key Management System (KMS). 

> When a KEK is rotated, the new key material is eventually used for newly produced records. Existing records, encrypted with older KEK versions, remain decryptable as long as the previous KEK versions are still in the KMS.

Reviewer: ```suggestion
When a KEK is rotated, the new key material is eventually used for newly produced records. Existing records, encrypted with older KEK versions, remain decryptable as long as the previous KEK versions are still available in the KMS.
```

Some of the KMS systems have a rather sophisticated view of the status of a KEK so it might still be in there but not accessible (at least thats how I understand the model)

> When the KEK is rotated in the external KMS, it will take up to 2 hours (by default) before all records produced by the filter

Reviewer: Isn't it `encryptionDekRefreshAfterWriteSeconds` that matters for encryptions and this the default is 1 hour?

Author: It's tricky since a DEK can be used right up until expiry if there are no accesses for a long time. Keith had the example like:
- 01:00 - DEK A created 
- 02:00 - DEK A becomes eligible for refresh
- 02:59 - user accesses DEK A triggering async refresh but gets DEK A from cache

If a DEK is being used frequently it should refresh close to 1h.

Author: For the overview I don't know whether to get into the details or just talk about the upper bound

Reviewer: I think the overview can assume topics with a regular flow of messages, as long as we document the need for message flow and how everything fits together in the extreme somewhere.

contain a DEK encrypted with the new key material. This is because existing encrypted DEKs are used for a configurable
amount of time after creation, the Filter caches the encrypted DEK.

> In a worst case scenario like the key material being compromised you can execute a rolling restart of your cluster of
> Kroxylicious instances after the key rotation. After the restart, messages being encrypted by the Record Encryption Filter
> will all use the latest key material.

Reviewer 2: (Non blocking) If kek material is compromised, the user got much more to be worried about (the encrypted messages in the transaction log).  In time, we'll need documentation etc to help a user cope, but I don't think we want to get in that now.

```suggestion
If you need to rotate key material immediately, execute a rolling restart of your cluster of Kroxylicious instances.
```
